src/test/kotlin/apply/domain/authenticationcode/AuthenticationCodeTest.kt

```diff
@@ -3,58 +3,53 @@ package apply.domain.authenticationcode
 import apply.EMAIL
 import apply.INVALID_CODE
 import apply.VALID_CODE
-import org.assertj.core.api.Assertions.assertThat
-import org.junit.jupiter.api.Test
-import org.junit.jupiter.api.assertAll
-import org.junit.jupiter.api.assertThrows
+import io.kotest.assertions.assertSoftly
+import io.kotest.assertions.throwables.shouldThrowExactly
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.booleans.shouldBeFalse
+import io.kotest.matchers.booleans.shouldBeTrue
+import io.kotest.matchers.nulls.shouldNotBeNull
 import java.time.LocalDateTime
 
-internal class AuthenticationCodeTest {
-    @Test
-    fun `인증 코드를 생성한다`() {
+class AuthenticationCodeTest : StringSpec({
+    "인증 코드를 생성한다" {
         val authenticationCode = AuthenticationCode(EMAIL)
-        assertAll(
-            { assertThat(authenticationCode.code).isNotNull() },
-            { assertThat(authenticationCode.authenticated).isFalse() },
-            { assertThat(authenticationCode.createdDateTime).isNotNull() }
-        )
+        assertSoftly {
+            authenticationCode.code.shouldNotBeNull()
+            authenticationCode.authenticated.shouldBeFalse()
+            authenticationCode.createdDateTime.shouldNotBeNull()
+        }
```

Comment:
와...이런 기능도 존재하는 군요!
공유 감사합니다 👍

Reply:
@NewWisdom 
전역으로 설정할 때 테스트 속도에 영향을 미치는지 궁금하네요. 별개로 전역적으로 설정하면 https://github.com/woowacourse/service-apply/pull/573#discussion_r926164638 의 이유로 `NullPointerException`을 볼 수 있습니다.

Reply:
@Rok93 
수신자를 지정하면 테스트 내에 같은 이름의 변수가 있으면 `this`를 추가해야 하며, 실수로 `this`를 추가하지 않아도 코드 리뷰를 통해 알아차리기 어려운 단점이 있습니다. 그렇지만 일단 반영하였어요!

Reply:
### 요약

- JUnit 5의 `assertAll`은 `NullPointerException`이 발생하더라도 끝까지 테스트를 실행한다.
- Kotest의 `assertSoftly`은 `NullPointerException` 등 예외가 발생하면 테스트를 중단한다.
   - `actual shouldNotBe null`은 테스트를 중단하지 않는다.
   - `actual shouldNot beNull()`은 테스트를 중단하지 않는다.
   - `actual.shouldNotBeNull()`은 `NullPointerException`이 발생해 테스트를 중단한다.

### Soft Assertions 작동 방식

1. `ErrorCollectionMode.Soft`가 아니면 `ErrorCollectionMode.Soft`로 설정하고 테스트를 실행한다.

```kotlin
// all.kt#L45-L58
inline fun <T> assertSoftly(assertions: () -> T): T {
   // Handle the edge case of nested calls to this function by only calling throwCollectedErrors in the
   // outermost verifyAll block
   if (errorCollector.getCollectionMode() == ErrorCollectionMode.Soft) return assertions()
   errorCollector.setCollectionMode(ErrorCollectionMode.Soft)
   return try {
      assertions()
   } finally {
      // In case if any exception is thrown from assertions block setting errorCollectionMode back to hard
      // so that it won't remain soft for others tests. See https://github.com/kotest/kotest/issues/1932
      errorCollector.setCollectionMode(ErrorCollectionMode.Hard)
      errorCollector.throwCollectedErrors()
   }
}
```

2. `shouldBe()`의 경우 `eq()`의 결과가 `null`이 아니면(=`AssertionError`이면) `collectOrThrow()`를 호출한다.

```kotlin
// should.kt#L16-L26
@Suppress("UNCHECKED_CAST")
infix fun <T, U : T> T.shouldBe(expected: U?) {
   when (expected) {
      is Matcher<*> -> should(expected as Matcher<T>)
      else -> {
         val actual = this
         assertionCounter.inc()
         eq(actual, expected)?.let(errorCollector::collectOrThrow)
      }
   }
}
```

3. 이전에 `ErrorCollectionMode.Soft`로 설정했기 때문에 `error`를 저장한다.

```kotlin
// ErrorCollector.kt#L123-L128
fun ErrorCollector.collectOrThrow(error: Throwable) {
   when (getCollectionMode()) {
      ErrorCollectionMode.Soft -> pushError(error)
      ErrorCollectionMode.Hard -> throw error
   }
}
```

4. `should()`의 경우에도 `collectOrThrow()`를 호출한다.

```kotlin
// should.kt#L41-L57
fun <T> invokeMatcher(t: T, matcher: Matcher<T>): T {
   assertionCounter.inc()
   val result = matcher.test(t)
   if (!result.passed()) {
      when (result) {
         is ComparableMatcherResult -> errorCollector.collectOrThrow(
            failure(
               Expected(Printed(result.expected())),
               Actual(Printed(result.actual())),
               result.failureMessage()
            )
         )
         else -> errorCollector.collectOrThrow(failure(result.failureMessage()))
      }
   }
   return t
}
```

5. `shouldNotBeNull()`의 경우 `null`이면 `error`가 저장되지만 마지막 `this!!` 때문에 예외가 발생한다.

```kotlin
// matches.kt#L67-L74
fun <T> T?.shouldNotBeNull(): T {
   contract {
      returns() implies (this@shouldNotBeNull != null)
   }

   this shouldNot beNull()
   return this!!
}
```

6. `assertSoftly`의 `try-finally`에서 발생한 예외는 폐기되고 이전에 저장한 `errors`만 표시된다.

```kotlin
// all.kt#L45-L58
inline fun <T> assertSoftly(assertions: () -> T): T {
   // Handle the edge case of nested calls to this function by only calling throwCollectedErrors in the
   // outermost verifyAll block
   if (errorCollector.getCollectionMode() == ErrorCollectionMode.Soft) return assertions()
   errorCollector.setCollectionMode(ErrorCollectionMode.Soft)
   return try {
      assertions()
   } finally {
      // In case if any exception is thrown from assertions block setting errorCollectionMode back to hard
      // so that it won't remain soft for others tests. See https://github.com/kotest/kotest/issues/1932
      errorCollector.setCollectionMode(ErrorCollectionMode.Hard)
      errorCollector.throwCollectedErrors()
   }
}
```